packages/internal/src/callback.js

```diff
@@ -155,3 +176,86 @@ export const isCallback = callback => {
   );
 };
 harden(isCallback);
+
+/**
+ * Prepare an attenuator class whose methods can be redirected via callbacks.
+ *
+ * @param {object} opts
+ * @param {InterfaceGuard} [opts.interfaceGuard] An interface guard for the new
+ * attenuator.
+ * @param {Iterable<PropertyKey>} [opts.methodNames] Methods to forward
+ * (default determined by interfaceGuard).
+ * @param {import('@agoric/zone').Zone} [opts.zone=heapZone] The zone in which to allocate attenuators.
+ */
+export const prepareAttenuator = ({
+  interfaceGuard,
+  methodNames,
+  zone = heapZone,
+}) => {
+  if (!methodNames) {
+    if (!interfaceGuard) {
+      throw Fail`interfaceGuard is required if methodKeys is missing`;
+    }
+    methodNames = Object.keys(interfaceGuard.methodGuards);
+  }
+
+  // Cache the names for callback verification.
+  const callbackKeys = new Set(methodNames);
+
+  /**
+   * @typedef {(this: any, ...args: unknown[]) => any} Method
+   */
+  /** @type {Record<any, Method>} */
+  const methods = {};
+  for (const key of callbackKeys) {
+    // Only allow the `PropertyKey` type for the target method key.
+    if (!isPropertyKey(key)) {
+      throw Fail`key ${q(key)} is not a PropertyKey`;
+    }
+
+    methods[key] = {
+      // Explicitly use concise method syntax to preserve `this` but prevent
+      // constructor behavior.
+      /** @type {Method} */
+      [key](...args) {
+        // Support both synchronous and async callbacks.
+        const cb = this.state.cbs[key];
+        cb || Fail`unimplemented method ${q(key)}`;
+        if (cb.isSync) {
+          return callSync(cb, ...args);
+        }
+        return callE(cb, ...args);
+      },
+    }[key];
+  }
+
+  /**
+   * Create an exo object whose behavior is composed from method names mapped
+   * to individual callbacks.  Nullish callbacks are ignored.
+   *
+   * @type {<Cbs extends Record<any, Callback<any> | null>>(callbacks: Cbs) => Farable<Record<keyof Cbs, (...args: unknown[]) => any>>}
+   */
+  const makeAttenuator = zone.exoClass(
+    'Attenuator',
+    interfaceGuard,
+    callbacks => {
+      /** @type {Record<any, Callback<any>>} */
+      const cbs = {};
+      for (const [key, cb] of Object.entries(callbacks)) {
+        if (cb == null) {
+          // Skip nullish.
+          continue;
+        }
+        callbackKeys.has(key) ||
+          Fail`Attenuator callbacks[${q(key)}] not allowed by methodKeys`;
+        isCallback(cb) ||
+          Fail`Attenuator callbacks[${q(key)}] is not a callback; got ${cb}`;
+        cbs[key] = cb;
+      }
+      return harden({ cbs });
```

Comment:
This does not seem to enforce that all declared `methodNames` are provided. Is that intended?